### PR TITLE
Add loader-v4 instruction constructors

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
Need convenient methods to construct loader-v4 instructions. These will be useful in CLI and other places.

#### Summary of Changes
Add the constructors.

Fixes #

### CODE REVIEW

**Comment by dmakarov** [comment]
LGTM. I assume there will be some tests that will use these methods?

**Comment by pgarg66** [comment]
> LGTM. I assume there will be some tests that will use these methods?

Yes, need to add some tests. I looked into `bpf_loader_upgradeable.rs` for some inspiration, but those tests are not as useful. `loader-v4/src/lib.rs` is another place where the tests could potentially be added.

**Comment by dmakarov** [comment]
> > LGTM. I assume there will be some tests that will use these methods?
> 
> Yes, need to add some tests. I looked into `bpf_loader_upgradeable.rs` for some inspiration, but those tests are not as useful. `loader-v4/src/lib.rs` is another place where the tests could potentially be added.

Sounds good. I thought maybe in some tests that construct instructions directly, these methods could be used, if that's convenient.

**Comment by pgarg66** [comment]
> > > LGTM. I assume there will be some tests that will use these methods?
> > 
> > 
> > Yes, need to add some tests. I looked into `bpf_loader_upgradeable.rs` for some inspiration, but those tests are not as useful. `loader-v4/src/lib.rs` is another place where the tests could potentially be added.
> 
> Sounds good. I thought maybe in some tests that construct instructions directly, these methods could be used, if that's convenient.

Yes, those should be simpler to add. Let me try and add it in this PR. I would also like to add some tests that calls the loader with the constructed instructions. Maybe that can go in a separate PR.

**Comment by pgarg66** [comment]
Added tests for the constructors.

**Review by dmakarov — DISMISSED** [review]
LGTM.

File: sdk/program/src/loader_v4.rs [hunk 1]
```diff
@@ -41,6 +46,130 @@ impl LoaderV4State {
...
+        id(),
+        &LoaderV4Instruction::Deploy,
+        vec![
+            AccountMeta::new(*program_address, false),
+            AccountMeta::new_readonly(*authority, true),
+        ],
+    )
+}
+
+/// Returns the instructions required to deploy a program.
```
Comment (Lichtso) [inline sdk/program/src/loader_v4.rs:126]:
```suggestion
/// Returns the instructions required to deploy a program using a buffer.
```

File: sdk/program/src/loader_v4.rs [hunk 2]
```diff
@@ -41,6 +46,130 @@ impl LoaderV4State {
...
+    !instruction_data.is_empty() && 4 == instruction_data[0]
+}
+
+/// Returns the instructions required to initialize a program/buffer account.
+pub fn create_buffer(
+    payer_address: &Pubkey,
+    buffer_address: &Pubkey,
+    lamports: u64,
+) -> Instruction {
+    system_instruction::create_account(payer_address, buffer_address, lamports, 0, &id())
```
Comment (Lichtso) [inline sdk/program/src/loader_v4.rs:75]: This should also take the `size` as a parameter and do the `Truncate` to the given `size` as well.

File: sdk/program/src/loader_v4.rs [hunk 3]
```diff
@@ -41,6 +46,130 @@ impl LoaderV4State {
...
+    program_address: &Pubkey,
+    authority: &Pubkey,
+    new_size: u32,
+    recipient_address: &Pubkey,
+) -> Instruction {
+    Instruction::new_with_bincode(
+        id(),
+        &LoaderV4Instruction::Truncate { new_size },
+        vec![
+            AccountMeta::new(*program_address, true),
```
Comment (Lichtso) [inline sdk/program/src/loader_v4.rs:89]: Here, the program does not need to sign. Only in the `Truncate` inside `create_buffer()` the program needs to be a signer to set the `authority` initially.

File: sdk/program/src/loader_v4.rs (hunk 2 above)
Comment (pgarg66) [reply sdk/program/src/loader_v4.rs:75]: So initial size will be 0, and truncate instruction will set the actual size?

Comment (Lichtso) [reply sdk/program/src/loader_v4.rs:75]: Yes, that is the idea. Truncate will also initialize the header struct for the program.

**Review by dmakarov — APPROVED** [review]

File: sdk/program/src/loader_v4.rs [hunk 4]
```diff
@@ -41,9 +46,147 @@ impl LoaderV4State {
...
+) -> Vec<Instruction> {
+    vec![
+        system_instruction::create_account(payer_address, buffer_address, lamports, 0, &id()),
+        Instruction::new_with_bincode(
+            id(),
+            &LoaderV4Instruction::Truncate { new_size },
+            vec![
+                AccountMeta::new(*buffer_address, true),
+                AccountMeta::new_readonly(*authority, true),
+                AccountMeta::new(*recipient_address, false),
```
Comment (Lichtso) [inline sdk/program/src/loader_v4.rs:86]: `recipient_address` is not needed and should be `payer_address` here as well.

Comment (pgarg66) [reply sdk/program/src/loader_v4.rs:86]: I thought about it. The caller can pass the payer address as the `recipient_address`.

**Pull request merged**